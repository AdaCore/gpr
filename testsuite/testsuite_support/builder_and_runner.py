import os
from e3.os.process import PIPE, Run, STDOUT
from random import getrandbits
from e3.testsuite.driver.classic import TestAbortWithFailure

# environment variables definition

USE_VALGRIND = "USE_VALGRIND"
USE_GNATCOV = "USE_GNATCOV"
COV_TRACES_DIR = "GNATCOV_TRACES_DIR"
COV_LEVEL = "GNATCOV_LEVEL"

GPR = os.environ["GPR2_TOOLS_PREFIX"] if "GPR2_TOOLS_PREFIX" in os.environ else "gpr"
GPRLS = GPR + "ls"
GPRCLEAN = GPR + "clean"
GPRINSTALL = GPR + "install"
GPRNAME = GPR + "name"
GPRCONFIG = GPR + "config"
GPRREMOTE = GPR + "remote"
GPRDOC = GPR + "doc"
GPRINSPECT = GPR + "inspect"


class BuilderAndRunner(object):
    """
    Handle program using gpr2 build. in coverage mode, instrument code &
    build instrumented code.

    Run program using gpr2 (including prebuilt gpr2-tools) using valgrind when
    requested.

    When using BuilderAndRunner() object is configured using python script
    parameters --valgrind --gnatcov --level --traces-dir

    In coverage mode, gpr2 programs's coverage traces are written in
    traces-dir using a random filename.

    This object provides wrappers for e3.os.process.Run, subprocess.call &
    subprocess.check_output
    """

    def __init__(self, driver=None):
        """configure BuilderAndRunner object using TestDriver environment or
        python script's argument when driver is None.
        """

        # associated TestDriver object
        self.driver = driver

        if driver is not None:

            # valgrind mode enabled status
            self.valgrind = driver.env.valgrind

            # coverage mode enabled status
            self.gnatcov = driver.env.gnatcov is not None

            if self.gnatcov:
                # gnatcov trace files directory
                self.traces_dir = driver.env.gnatcov.traces_dir

                # gnatcov calls --level switch
                self.level = driver.env.gnatcov.covlevel
            else:
                self.level = None
                self.traces_dir = None
        else:
            self.valgrind = USE_VALGRIND in os.environ
            self.gnatcov = USE_GNATCOV in os.environ
            self.traces_dir = os.environ.get(COV_TRACES_DIR)
            self.level = os.environ.get(COV_LEVEL)

    def simple_run(
        self,
        cmd,
        env=None,
        catch_error=True,
        output=PIPE,
        error=STDOUT,
        analyze_output=True,
    ):
        """ generic TestDriver.shell or e3.os.process.Run runner"""
        if self.driver is not None:
            if env is not None:
                effective_env = dict(os.environ)
                effective_env.update(env)
            return self.driver.shell(
                cmd,
                env=effective_env,
                catch_error=catch_error,
                analyze_output=analyze_output,
            )
        else:
            p = Run(cmd, env=env, output=output, error=error, ignore_environ=False)
            if catch_error and p.status != 0:
                print(str(cmd) + " returned " + str(p.status))
                if p.out:
                    print("stdout\n" + p.out)
                if p.err:
                    print("stderr\n" + p.err)
                raise TestAbortWithFailure("non-zero exit status")
            else:
                return p

    def build(self, project, vars=[], args=[], env=None, output=PIPE):
        """ gprbuild wrapper for normal & coverage modes """
        # If code coverage is requested, leave a chance to gnatcov to decorate
        # the execution of the subprogram in order to make it contribute to
        # code coverage.
        if self.gnatcov:
            gnatcov_cmd = [
                "gnatcov",
                "instrument",
                "--level",
                self.level,
                "--dump-trigger=atexit",
                "--externally-built-projects",
                "--projects",
                "gpr2",
                "--no-subprojects",
                "-P",
                project,
            ] + vars
            self.simple_run(gnatcov_cmd, env=env, analyze_output=False)

            # to do coverage, we use the repo's gpr2 project file instead of
            # the installed one from libgpr2. This means we need to ensure
            # that some scenario variables are properly set.
            gprbuild_cmd = (
                ["gprbuild", "-P", project,
                 "-XGPR2_BUILD=gnatcov", "-XXMLADA_BUILD=static"]
                + vars
                + ["--src-subdirs=gnatcov-instr", "--implicit-with=gnatcov_rts"]
                + args
            )
        else:
            gprbuild_cmd = ["gprbuild", "-P", project] + vars + args

        return self.simple_run(gprbuild_cmd, env=env, output=output)

    def run(self, cmd, env=None, output=PIPE, catch_error=False):
        """ generic wrapper handling coverage & valgrind modes. """
        if self.gnatcov:
            if env is None:
                env = {}
            file = str(getrandbits(128)) + ".srctrace"
            env["GNATCOV_TRACE_FILE"] = os.path.join(self.traces_dir, file)
        run_cmd = (["valgrind", "-q"] if self.valgrind else []) + cmd

        return self.simple_run(run_cmd, env=env, catch_error=catch_error, output=output)

    def check_output(self, cmd):
        """subprocess.check_output wrapper handling coverage & valgrind
        modes.
        """
        return self.run(cmd, catch_error=True)

    def call(self, cmd):
        """subprocess.call wrapper handling coverage & valgrind
        modes.
        """
        p = self.run(cmd)
        print(p.out, end="")
        return p

    def check_call(self, cmd):
        """subprocess.call wrapper handling coverage & valgrind
        modes.
        """
        p = self.run(cmd, catch_error=True)
        print(p.out, end="")
        return p

    def insert_build_and_runner_parameters(self, env):
        """The BuildAndRunner initialization parameters to allow a
        BuilderAndRunner() call in the python test script.
        """
        if self.valgrind:
            env[USE_VALGRIND] = "true"
        if self.gnatcov:
            env[USE_GNATCOV] = "true"
            env[COV_TRACES_DIR] = self.traces_dir
            env[COV_LEVEL] = self.level
