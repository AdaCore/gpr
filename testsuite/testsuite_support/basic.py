import os
import re
from e3.testsuite.driver.classic import TestAbortWithError
from e3.testsuite.driver.classic import ClassicTestDriver
from e3.fs import cp
from testsuite_support.base_driver import create_fake_ada_compiler
from testsuite_support.builder_and_runner import BuilderAndRunner

TESTSUITE_ROOT_DIR = os.path.dirname(os.path.dirname(os.path.abspath(__file__)))
ADA_SUPPORT_FILES = os.path.join(TESTSUITE_ROOT_DIR, "support")


class BasicDriver(ClassicTestDriver):
    """
    Driver to build a test program and run it.

    Interface:

    * put a main in test.adb that will be compiled (using a default project)
    * use Test_Assert functions to make your assertions
    * return Test_Assert.Report at the end of our main procedure

    This driver will build the program using GPRbuild and will then run the
    program and ensure that all assertions did passed and that the program
    did not crashed
    """

    def set_up(self):
        self.project_file = self.test_env.get("project_file")
        if self.project_file is None:
            cp(
                os.path.join(ADA_SUPPORT_FILES, "test.gpr"),
                self.test_env["working_dir"],
            )
            self.project_file = "./test.gpr"
            self.main = "./obj/test"
        else:
            self.main = self.test_env.get("./main")

        if not self.project_file or not isinstance(self.project_file, str):
            raise TestAbortWithError(
                'test.yaml: please define a "project_file" string field'
            )
        if not self.main or not isinstance(self.main, str):
            raise TestAbortWithError('test.yaml: please define a "main" string field')

        self.fake_ada_target = self.test_env.get("fake_ada_target")
        self.builder_and_runner = BuilderAndRunner(self)

    def run(self):
        env = {"ADA_SUPPORT_SOURCE_DIR": ADA_SUPPORT_FILES}

        # If we are requested to run with a fake toolchain, set it up now
        if self.fake_ada_target:
            fake_dir = self.working_dir("fake-ada")
            create_fake_ada_compiler(
                self,
                comp_dir=fake_dir,
                comp_target=self.fake_ada_target,
                gnat_version="21.0w",
                gcc_version="8.4.3",
                runtimes=["rtp"],
                comp_is_cross=True,
            )
            env["PATH"] = os.path.join(fake_dir, "bin") + os.path.pathsep + env["PATH"]

        # Build the program and run it
        p = self.builder_and_runner.build(
            project=self.project_file,
            args=["-g1", "-p", "-bargs", "-Es"],
            env=env,
        )
        self.output += p.out
        p = self.builder_and_runner.run([self.main], env=env)
        self.output += p.out
        if "Call stack traceback locations" in p.out:
            m = re.search(r"Call stack traceback locations:.*\n([^\r\n]+)", p.out)
            if m.group(1) is not None:
                self.shell(["addr2line", "-e", self.main] + m.group(1).split())

    def compute_failures(self):
        return (
            ["Success marker not found"]
            if "<=== TEST PASSED ===>" not in self.result.log.log
            else []
        )
