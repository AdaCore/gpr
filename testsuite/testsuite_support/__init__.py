import e3.testsuite

from .build_and_run import BuildAndRunDriver
from .basic import BasicDriver
from testsuite_support.gnatcov import GNATcov
from testsuite_support.python_script import PythonScriptDriver


class Testsuite(e3.testsuite.Testsuite):
    tests_subdir = 'tests'
    enable_cross_support = True
    test_driver_map = {
        'build_and_run': BuildAndRunDriver,
        'python_script': PythonScriptDriver,
        'basic': BasicDriver
    }

    @property
    def auto_generate_text_report(self) -> bool:
        "Generate a text report together with the raw results file"
        return True

    def add_options(self, parser):
        parser.add_argument(
            "--valgrind",
            action="store_true",
            help="Run test executable under valgrind.")

        parser.add_argument(
            "--from-gnat",
            action="store_true",
            help="If provided, means gnat packages are available")

        parser.add_argument(
            "--gnatcov",
            help="If provided, compute the source code coverage of testcases"
                 " on ALS. This requires GNATcoverage working with"
                 " instrumentation. The argument passed must be a directory"
                 " that contains all SID files.")

        parser.add_argument(
            "--level",
            help="coverage level. can be:"
                 " branch, insn, stmt, stmt+decision, stmt+mcdc, stmt+uc_mcdc")

    def set_up(self):
        super(Testsuite, self).set_up()
        self.env.valgrind = self.main.args.valgrind
        self.env.from_gnat = self.main.args.from_gnat

        # If code coverage is requested, initialize our helper and build
        # instrumented programs.
        self.env.gnatcov = (GNATcov(self) if self.main.args.gnatcov else None)
        self.env.root_dir = self.root_dir

    def tear_down(self):
        if self.env.gnatcov:
            self.env.gnatcov.report()

        super(Testsuite, self).tear_down()
