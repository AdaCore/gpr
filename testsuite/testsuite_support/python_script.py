import os
import sys

from testsuite_support.base_driver import BaseDriver
from testsuite_support.builder_and_runner import BuilderAndRunner


class PythonScriptDriver(BaseDriver):
    """
    Driver to run a Python script.

    Interface:

    * put a "test.py" script in the test directory;
    * put a "test.out" text file in the test directory.

    This driver will run the Python script. Its output is then checked against
    the expected output (test.out file). This mechanism is the most flexible
    way to write a testcase, but also the more verbose one and the most complex
    one. Use this driver when no other one fits.
    """

    def run(self):
        builder_and_runner = BuilderAndRunner(self)

        cmd = [sys.executable, 'test.py']

        env = dict(os.environ)
        if 'PYTHONPATH' not in os.environ:
            env['PYTHONPATH'] = self.env.root_dir
        else:
            env['PYTHONPATH'] = str(self.env.root_dir) + \
                                os.path.pathsep + os.environ['PYTHONPATH']
        builder_and_runner.insert_build_and_runner_parameters(env)

        env["root_dir"] = self.env.root_dir

        self.shell(cmd, env=env)
