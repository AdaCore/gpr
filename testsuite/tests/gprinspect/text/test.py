import os
import re
from testsuite_support.builder_and_runner import BuilderAndRunner, GPRINSPECT

bnr = BuilderAndRunner()

p = bnr.run([GPRINSPECT, "-Proot.gpr", "--all", "-r"])

output = p.out

# Make sure build-specific values are normalized for test output comparison
# purpose
output = re.sub(r"(Generated on *:) .*", r"\1 2000-01-01 00:00:00", output)
output = re.sub(r"(Version *:) .*", r"\1 Pro 1.0 (20000101) (<host>)", output)
output = output.replace(os.getcwd(), "<cwd>")

# Windows paths encoded in view_ids:
# in view_ids, all paths are lower-cased (case insensitive filesystem)
if os.name == 'nt':
    cwd = os.getcwd().lower()
    # first replace $PWD with a generic value (backslashes as path separators on
    # both sides)
    output = output.replace(cwd, "<cwd>")
    # for the other paths, replace backslashes with slashes
    output = output.replace("\\", "/")
    # check windows drive, and remove it to have "root" paths only
    output = re.sub(r"(.*)[a-zA-Z]://", r"\1/", output)

# now for all hosts, make sure the runtime path is normalized for test
# output comparison purpose
output = re.sub(r"( *- ).*/ada(lib|include)(.*)", r"\1<gnat rts>/ada\2\3", output)
out = []
in_path = False

for line in output.splitlines():
    if line.startswith("* Project search paths"):
        out.append(line)
        in_path = True
        out.append("    - <filtered-out>")
    elif in_path and line.startswith("* "):
        in_path = False
    if not in_path:
        out.append(line)

output = '\n'.join(out)
print(output)
