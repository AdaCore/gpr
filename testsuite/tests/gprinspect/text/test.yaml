description:
    check text output of gprinspect
driver: python_script
output:
- ["test.out.win", "'windows' in env.host.platform"]
