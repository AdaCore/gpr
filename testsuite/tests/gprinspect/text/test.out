+--------------------------------------+
|         General Information          |
+--------------------------------------+
* Generated on : 2000-01-01 00:00:00
* Version      : Pro 1.0 (20000101) (<host>)

+--------------------------------------+
|       Project Tree Information       |
+--------------------------------------+
* Messages :
    - lib.gpr: lint: project qualifier could be explicitly set to library project
* Project count        :  6
* Project search paths :
    - <filtered-out>
* Object search paths  :
    - <cwd>/
    - <cwd>/obj/
    - <cwd>/obj2/
    - <gnat rts>/adalib/
* Source search paths  :
    - <cwd>
    - <cwd>/src
    - <cwd>/src2
    - <gnat rts>/adainclude/
* Root project :
    - Root (<<cwd>/root.gpr)

+--------------------------------------+
|         Projects Information         |
+--------------------------------------+
* Root (<<cwd>/root.gpr) [ aggregate project ]
    - Project file      : <cwd>/root.gpr
    - Project directory : <cwd>/
    - Aggregated        : 
       - Agg ($<cwd>/agg.gpr)
    - Attributes        : 
       - Externally_Built [ SINGLE ]
          - Value : "false"
       - Name [ SINGLE ]
          - Value : "root"
       - Object_Dir [ SINGLE ]
          - Value : "."
       - Project_Dir [ SINGLE ]
          - Value : "<cwd>/"
       - Project_Files [ LIST ]
          - Values : 
             - "agg.gpr"

* Agg ($<cwd>/agg.gpr) [ aggregate project ]
    - Project file      : <cwd>/agg.gpr
    - Project directory : <cwd>/
    - Aggregated        : 
       - Prj ($<cwd>/prj.gpr)
    - Attributes        : 
       - Externally_Built [ SINGLE ]
          - Value : "false"
       - Name [ SINGLE ]
          - Value : "agg"
       - Object_Dir [ SINGLE ]
          - Value : "."
       - Project_Dir [ SINGLE ]
          - Value : "<cwd>/"
       - Project_Files [ LIST ]
          - Values : 
             - "prj.gpr"
    - Variables      : 
       - Answer [ SINGLE ]
          - Variable type : "Yes_No"
          - Value : "yes"
       - Var2 [ SINGLE ]
          - Value : "no"
    - Types          : 
       - Yes_No
          - Values : 
             - "yes"
             - "no"

* Shared ($<cwd>/shared.gpr>$<cwd>/prj.gpr) [ abstract project ]
    - Project file      : <cwd>/shared.gpr
    - Project directory : <cwd>/
    - Extended by       : Prj ($<cwd>/prj.gpr)
    - Imported-by        : 
       - Lib ($<cwd>/lib.gpr>$<cwd>/prj.gpr)
       - Base ($<cwd>/base.gpr>$<cwd>/prj.gpr)
    - Attributes        : 
       - Exec_Dir [ SINGLE ]
          - Value : "."
       - Externally_Built [ SINGLE ]
          - Value : "false"
       - Library_Auto_Init [ SINGLE ]
          - Value : "false"
       - Library_Kind [ SINGLE ]
          - Value : "static"
       - Library_Standalone [ SINGLE ]
          - Value : "no"
       - Name [ SINGLE ]
          - Value : "shared"
       - Object_Dir [ SINGLE ]
          - Value : "."
       - Project_Dir [ SINGLE ]
          - Value : "<cwd>/"
    - Variables      : 
       - Languages [ LIST ]
          - Values : 
             - "Ada"
             - "C"

* Lib ($<cwd>/lib.gpr>$<cwd>/prj.gpr) [ library project ]
    - Project file      : <cwd>/lib.gpr
    - Project directory : <cwd>/
    - Object directory  : <cwd>/
    - Source directory  :
       - <cwd>
    - Library name      : somename
    - Library file      : <cwd>/lib/libsomename.a
    - Library directory : <cwd>/lib/
    - Library ALI dir.  : <cwd>/lib/
    - Extended by       : Prj ($<cwd>/prj.gpr)
    - Imported-by        : 
       - Base ($<cwd>/base.gpr>$<cwd>/prj.gpr)
    - Imports           : 
       - Shared ($<cwd>/shared.gpr>$<cwd>/prj.gpr)
    - Attributes        : 
       - Exec_Dir [ SINGLE ]
          - Value : "."
       - Externally_Built [ SINGLE ]
          - Value : "false"
       - Languages [ LIST ]
          - Values : 
             - "Ada"
       - Library_Ali_Dir [ SINGLE ]
          - Value : "lib"
       - Library_Auto_Init [ SINGLE ]
          - Value : "false"
       - Library_Dir [ SINGLE ]
          - Value : "lib"
       - Library_Kind [ SINGLE ]
          - Value : "static"
       - Library_Name [ SINGLE ]
          - Value : "somename"
       - Library_Src_Dir [ SINGLE ]
          - Value : "lib"
       - Library_Standalone [ SINGLE ]
          - Value : "no"
       - Name [ SINGLE ]
          - Value : "lib"
       - Object_Dir [ SINGLE ]
          - Value : "."
       - Project_Dir [ SINGLE ]
          - Value : "<cwd>/"
       - Source_Dirs [ LIST ]
          - Values : 
             - "."

* Base ($<cwd>/base.gpr>$<cwd>/prj.gpr) [ standard project ]
    - Project file      : <cwd>/base.gpr
    - Project directory : <cwd>/
    - Object directory  : <cwd>/obj/
    - Source directory  :
       - <cwd>/src
    - Extended by       : Prj ($<cwd>/prj.gpr)
    - Imports           : 
       - Lib ($<cwd>/lib.gpr>$<cwd>/prj.gpr)
       - Shared ($<cwd>/shared.gpr>$<cwd>/prj.gpr)
    - Attributes        : 
       - Exec_Dir [ SINGLE ]
          - Value : "obj"
       - Externally_Built [ SINGLE ]
          - Value : "false"
       - Languages [ LIST ]
          - Values : 
             - "Ada"
       - Name [ SINGLE ]
          - Value : "base"
       - Object_Dir [ SINGLE ]
          - Value : "obj"
       - Project_Dir [ SINGLE ]
          - Value : "<cwd>/"
       - Source_Dirs [ LIST ]
          - Values : 
             - "src"

* Prj ($<cwd>/prj.gpr) [ standard project ]
    - Project file      : <cwd>/prj.gpr
    - Project directory : <cwd>/
    - Object directory  : <cwd>/obj2/
    - Source directory  :
       - <cwd>/src2
    - Extends           : 
       - Base ($<cwd>/base.gpr>$<cwd>/prj.gpr)
       - Lib ($<cwd>/lib.gpr>$<cwd>/prj.gpr)
       - Shared ($<cwd>/shared.gpr>$<cwd>/prj.gpr)
    - Attributes        : 
       - Exec_Dir [ SINGLE ]
          - Value : "obj2"
       - Externally_Built [ SINGLE ]
          - Value : "false"
       - Languages [ LIST ]
          - Values : 
             - "Ada"
       - Name [ SINGLE ]
          - Value : "prj"
       - Object_Dir [ SINGLE ]
          - Value : "obj2"
       - Project_Dir [ SINGLE ]
          - Value : "<cwd>/"
       - Source_Dirs [ LIST ]
          - Values : 
             - "src2"
       - Naming'Body_Suffix [ SINGLE ]
          - Index value : "ada"
          - Value : ".adb"
       - Naming'Body_Suffix [ SINGLE ]
          - Index value : "c"
          - Value : ".c"
       - Naming'Dot_Replacement [ SINGLE ]
          - Value : "~"
       - Naming'Implementation_Suffix [ SINGLE ]
          - Index value : "ada"
          - Value : ".adb"
       - Naming'Implementation_Suffix [ SINGLE ]
          - Index value : "c"
          - Value : ".c"
       - Naming'Spec_Suffix [ SINGLE ]
          - Index value : "ada"
          - Value : ".ads"
       - Naming'Spec_Suffix [ SINGLE ]
          - Index value : "c"
          - Value : ".h"
       - Naming'Specification_Suffix [ SINGLE ]
          - Index value : "ada"
          - Value : ".ads"
       - Naming'Specification_Suffix [ SINGLE ]
          - Index value : "c"
          - Value : ".h"

{
  "info": {
    "generated-on": 2000-01-01 00:00:00
    "version": Pro 1.0 (20000101) (<host>)
  },
  "tree": {
    "messages": [
      "lib.gpr: lint: project qualifier could be explicitly set to library project"
    ],
    "stats": {
      "project-count": 3
    },
    "project-search-paths": [
      <filtered-out>
    ],
    "object-search-paths": [
      "<gnat rts>/adalib/",
      "<cwd>/obj2/"
    ],
    "source-search-paths": [
      "<gnat rts>/adainclude/",
      "<cwd>/src2"
    ],
    "root-project": {
      "name": "Root",
      "id": "<<cwd>/root.gpr"
    }
  },
  "projects": [
    {
      "project": {
        "id": "$<cwd>/prj.gpr",
        "name": "Prj",
        "kind": "standard",
        "qualifier": "standard",
        "simple-name": "prj.gpr",
        "file-name": "<cwd>/prj.gpr",
        "directory": "<cwd>/",
        "object-directory": "<cwd>/obj2/",
        "source-directories": [
          "<cwd>/src2"
        ]
      },
      "extended": [
        "$<cwd>/base.gpr>$<cwd>/prj.gpr",
        "$<cwd>/lib.gpr>$<cwd>/prj.gpr",
        "$<cwd>/shared.gpr>$<cwd>/prj.gpr"
      ],
      "imported-by": [
        "$<cwd>/agg.gpr"
      ],
      "packages": [
        {
          "name": "Binder",
          "attributes": [
            {
              "name": "Bindfile_Option_Substitution",
              "index": "-shared",
              "kind": "list",
              "values": [
                "-shared-libgcc"
              ]
            },
            {
              "name": "Bindfile_Option_Substitution",
              "index": "-static",
              "kind": "list",
              "values": [
                "-static-libgcc"
              ]
            },
            {
              "name": "Driver",
              "index": "Ada",
              "kind": "single",
              "value": "<root>/gprbuild/gprbind/"
            },
            {
              "name": "Objects_Path_File",
              "index": "Ada",
              "kind": "single",
              "value": "ADA_PRJ_OBJECTS_FILE"
            },
            {
              "name": "Required_Switches",
              "index": "Ada",
              "kind": "list",
              "values": [
                "gnatbind_prefix="
              ]
            }
          ]
        },
        {
          "name": "Clean",
          "attributes": [
            {
              "name": "Object_Artifact_Extensions",
              "index": "Ada",
              "kind": "list",
              "values": [
                ".s",
                ".ci",
                ".gcno",
                ".su"
              ]
            },
            {
              "name": "Object_Artifact_Extensions",
              "index": "C",
              "kind": "list",
              "values": [
                ".s",
                "ci",
                ".gcno",
                ".su"
              ]
            },
            {
              "name": "Object_Artifact_Extensions",
              "index": "C++",
              "kind": "list",
              "values": [
                ".s",
                "ci",
                ".gcno",
                ".su"
              ]
            },
            {
              "name": "Source_Artifact_Extensions",
              "index": "Ada",
              "kind": "list",
              "values": [
                ".dg",
                ".rep"
              ]
            },
            {
              "name": "Source_Artifact_Extensions",
              "index": "C",
              "kind": "list",
              "values": [
                ".gli"
              ]
            },
            {
              "name": "Source_Artifact_Extensions",
              "index": "C++",
              "kind": "list",
              "values": [
                ".gli"
              ]
            }
          ]
        },
        {
          "name": "Compiler",
          "attributes": [
            {
              "name": "Driver",
              "index": "Ada",
              "kind": "single",
              "value": "<root>/bin/gcc/"
            },
            {
              "name": "Config_Body_File_Name",
              "index": "Ada",
              "kind": "single",
              "value": "pragma Source_File_Name_Project (%u, Body_File_Name => /"%f/");"
            },
            {
              "name": "Config_Body_File_Name_Index",
              "index": "Ada",
              "kind": "single",
              "value": "pragma Source_File_Name_Project (%u, Body_File_Name => /"%f/", Index => %i);"
            },
            {
              "name": "Config_Body_File_Name_Pattern",
              "index": "Ada",
              "kind": "single",
              "value": "pragma Source_File_Name_Project   (Body_File_Name  => /"*%b/",   Casing          => %c,   Dot_Replacement => /"%d/");"
            },
            {
              "name": "Config_File_Switches",
              "index": "Ada",
              "kind": "list",
              "values": [
                "-gnatec="
              ]
            },
            {
              "name": "Config_File_Unique",
              "index": "Ada",
              "kind": "single",
              "value": "False"
            },
            {
              "name": "Config_Spec_File_Name",
              "index": "Ada",
              "kind": "single",
              "value": "pragma Source_File_Name_Project (%u, Spec_File_Name => /"%f/");"
            },
            {
              "name": "Config_Spec_File_Name_Index",
              "index": "Ada",
              "kind": "single",
              "value": "pragma Source_File_Name_Project (%u, Spec_File_Name => /"%f/", Index => %i);"
            },
            {
              "name": "Config_Spec_File_Name_Pattern",
              "index": "Ada",
              "kind": "single",
              "value": "pragma Source_File_Name_Project   (Spec_File_Name  => /"*%s/",   Casing          => %c,   Dot_Replacement => /"%d/");"
            },
            {
              "name": "Dependency_Kind",
              "index": "Ada",
              "kind": "single",
              "value": "ALI_File"
            },
            {
              "name": "Include_Path_File",
              "index": "Ada",
              "kind": "single",
              "value": "ADA_PRJ_INCLUDE_FILE"
            },
            {
              "name": "Language_Kind",
              "index": "Ada",
              "kind": "single",
              "value": "unit_based"
            },
            {
              "name": "Leading_Required_Switches",
              "index": "Ada",
              "kind": "list",
              "values": [
                "-c",
                "-x",
                "ada",
                "-gnatA"
              ]
            },
            {
              "name": "Mapping_Body_Suffix",
              "index": "Ada",
              "kind": "single",
              "value": "%b"
            },
            {
              "name": "Mapping_File_Switches",
              "index": "Ada",
              "kind": "list",
              "values": [
                "-gnatem="
              ]
            },
            {
              "name": "Mapping_Spec_Suffix",
              "index": "Ada",
              "kind": "single",
              "value": "%s"
            },
            {
              "name": "Multi_Unit_Object_Separator",
              "index": "Ada",
              "kind": "single",
              "value": "~"
            },
            {
              "name": "Multi_Unit_Switches",
              "index": "Ada",
              "kind": "list",
              "values": [
                "-gnateI"
              ]
            },
            {
              "name": "Pic_Option",
              "index": "Ada",
              "kind": "list",
              "values": [
                "-fPIC"
              ]
            }
          ]
        },
        {
          "name": "Linker",
          "attributes": [
          ]
        },
        {
          "name": "Naming",
          "attributes": [
            {
              "name": "Body_Suffix",
              "index": "Ada",
              "kind": "single",
              "value": ".adb"
            },
            {
              "name": "Body_Suffix",
              "index": "Asm",
              "kind": "single",
              "value": ".s"
            },
            {
              "name": "Body_Suffix",
              "index": "Asm2",
              "kind": "single",
              "value": ".asm"
            },
            {
              "name": "Body_Suffix",
              "index": "Asm_Cpp",
              "kind": "single",
              "value": ".S"
            },
            {
              "name": "Body_Suffix",
              "index": "C",
              "kind": "single",
              "value": ".c"
            },
            {
              "name": "Body_Suffix",
              "index": "C++",
              "kind": "single",
              "value": ".cpp"
            },
            {
              "name": "Body_Suffix",
              "index": "Fortran",
              "kind": "single",
              "value": ".f"
            },
            {
              "name": "Dot_Replacement",
              "kind": "single",
              "value": "~"
            },
            {
              "name": "Spec_Suffix",
              "index": "Ada",
              "kind": "single",
              "value": ".ads"
            },
            {
              "name": "Spec_Suffix",
              "index": "C",
              "kind": "single",
              "value": ".h"
            },
            {
              "name": "Spec_Suffix",
              "index": "C++",
              "kind": "single",
              "value": ".hh"
            }
          ]
        }
      ],
      "attributes": [
        {
          "name": "Exec_Dir",
          "kind": "single",
          "value": "obj2"
        },
        {
          "name": "Externally_Built",
          "kind": "single",
          "value": "false"
        },
        {
          "name": "Inherit_Source_Path",
          "index": "Asm_Cpp",
          "kind": "list",
          "values": [
            "Asm"
          ]
        },
        {
          "name": "Inherit_Source_Path",
          "index": "C++",
          "kind": "list",
          "values": [
            "C"
          ]
        },
        {
          "name": "Languages",
          "kind": "list",
          "values": [
            "Ada"
          ]
        },
        {
          "name": "Name",
          "kind": "single",
          "value": "prj"
        },
        {
          "name": "Object_Dir",
          "kind": "single",
          "value": "obj2"
        },
        {
          "name": "Project_Dir",
          "kind": "single",
          "value": "<cwd>/"
        },
        {
          "name": "Runtime_Dir",
          "index": "Ada",
          "kind": "single",
          "value": "<root>/lib/gcc/"
        },
        {
          "name": "Source_Dirs",
          "kind": "list",
          "values": [
            "src2"
          ]
        },
        {
          "name": "Toolchain_Version",
          "index": "Ada",
          "kind": "single",
          "value": "GNAT <version>
        }
      ]
    },
    {
      "project": {
        "id": "$<cwd>/agg.gpr",
        "name": "Agg",
        "kind": "aggregate",
        "qualifier": "aggregate",
        "simple-name": "agg.gpr",
        "file-name": "<cwd>/agg.gpr",
        "directory": "<cwd>/"
      },
      "imported-by": [
        "<<cwd>/root.gpr"
      ],
      "aggregated": [
        "$<cwd>/prj.gpr"
      ],
      "attributes": [
        {
          "name": "Externally_Built",
          "kind": "single",
          "value": "false"
        },
        {
          "name": "Inherit_Source_Path",
          "index": "Asm_Cpp",
          "kind": "list",
          "values": [
            "Asm"
          ]
        },
        {
          "name": "Inherit_Source_Path",
          "index": "C++",
          "kind": "list",
          "values": [
            "C"
          ]
        },
        {
          "name": "Name",
          "kind": "single",
          "value": "agg"
        },
        {
          "name": "Object_Dir",
          "kind": "single",
          "value": "."
        },
        {
          "name": "Project_Dir",
          "kind": "single",
          "value": "<cwd>/"
        },
        {
          "name": "Project_Files",
          "kind": "list",
          "values": [
            "prj.gpr"
          ]
        },
        {
          "name": "Runtime_Dir",
          "index": "Ada",
          "kind": "single",
          "value": "<root>/lib/gcc/"
        },
        {
          "name": "Toolchain_Version",
          "index": "Ada",
          "kind": "single",
          "value": "GNAT <version>
        }
      ],
      "variables": [
        {
          "name": "Answer",
          "type": "Yes_No",
          "kind": "single",
          "value": "yes"
        },
        {
          "name": "Var2",
          "kind": "single",
          "value": "no"
        }
      ],
      "types": [
        {
          "name": "Yes_No",
          "values": [
            "yes",
            "no"
          ]
        }
      ]
    },
    {
      "project": {
        "id": "<<cwd>/root.gpr",
        "name": "Root",
        "kind": "aggregate",
        "qualifier": "aggregate",
        "simple-name": "root.gpr",
        "file-name": "<cwd>/root.gpr",
        "directory": "<cwd>/"
      },
      "aggregated": [
        "$<cwd>/agg.gpr"
      ],
      "attributes": [
        {
          "name": "Externally_Built",
          "kind": "single",
          "value": "false"
        },
        {
          "name": "Inherit_Source_Path",
          "index": "Asm_Cpp",
          "kind": "list",
          "values": [
            "Asm"
          ]
        },
        {
          "name": "Inherit_Source_Path",
          "index": "C++",
          "kind": "list",
          "values": [
            "C"
          ]
        },
        {
          "name": "Name",
          "kind": "single",
          "value": "root"
        },
        {
          "name": "Object_Dir",
          "kind": "single",
          "value": "."
        },
        {
          "name": "Project_Dir",
          "kind": "single",
          "value": "<cwd>/"
        },
        {
          "name": "Project_Files",
          "kind": "list",
          "values": [
            "agg.gpr"
          ]
        },
        {
          "name": "Runtime_Dir",
          "index": "Ada",
          "kind": "single",
          "value": "<root>/lib/gcc/"
        },
        {
          "name": "Toolchain_Version",
          "index": "Ada",
          "kind": "single",
          "value": "GNAT <version>
        }
      ]
    }
  ]
}
